libraries/core_libs/consensus/src/transaction/transaction_manager.cpp

```diff
@@ -21,6 +21,28 @@ TransactionManager::TransactionManager(FullNodeConfig const &conf, std::shared_p
   }
 }
 
+uint64_t TransactionManager::estimateTransactionByHash(const trx_hash_t &hash,
+                                                       std::optional<uint64_t> proposal_period) const {
+  const auto &trx = getTransaction(hash);
+  return estimateTransaction(trx, proposal_period);
+}
+
+uint64_t TransactionManager::estimateTransaction(std::shared_ptr<Transaction> trx,
+                                                 std::optional<uint64_t> proposal_period) const {
+  const auto &result = final_chain_->call(
+      state_api::EVMTransaction{
+          trx->getSender(),
+          trx->getGasPrice(),
+          trx->getReceiver(),
+          trx->getNonce(),
+          trx->getValue(),
+          trx->getGas(),
+          trx->getData(),
+      },
+      proposal_period);
```

Comment:
Since PBFT chain is async with execution of final chain. When final chain execution much behind to PBFT chain. That should not work here.

Reply:
In our experience, the proposal_period should always behind the finalized period. I think should be fine here.
Let's add "try" and "catch" for exception for final_chain_->call

Reply:
I didn't get what are you talking about in 1 message
Isn't it right that `proposal_period` should be always behind finalized period?

Reply:
so why should be proposal_period always behind finalized_period ?

Reply:
https://github.com/Taraxa-project/taraxa-node/pull/1672#discussion_r867748891